meetings/working-groups/params-improvements/PI-2022-11-03.md

```diff
@@ -0,0 +1,126 @@
+
+## Span reuse
+The compiler _may reuse a span_ when:
+- the span is implicitly allocated (as a `params` argument or as a _collection literal_),
+- the span length is known at compile time,
+- the compiler can determine _through escape analysis_ that no references to the span escape the containing method, and
+- the compiler can determine there are _no aliases_ to the span in user code.
+
+For arguments to a `params ReadOnlySpan<T>` parameter (which is implicitly `scoped`), the conditions above are satisfied.
+
+The compiler will reuse a single span across all calls to `params ReadOnlySpan<T>` methods where the span element types are considered identical by the runtime.
+
+The length of the reused span will be the length of the longest `params` argument list from all uses.
+(The actual span argument passed to a `params` method will be a slice of the reused span, with the expected length for the call.)
+
+Reuse may be across distinct call sites _or_ repeated calls from the same call site.
+Reuse is per thread of execution and within the same method only.
+No reuse across lambda expressions, local functions, and the containing methods.
+
+The reused span will be cleared after each use.
+
+To opt out of reuse at a call site, the calling code should allocate the span explicitly.
+
+_Should we only reuse spans that are allocated on the stack? If we also allow reuse of heap allocated buffers, that will require completely different code gen for managing allocations._
+
+## Collection literals
+We will support stack allocation and reuse of spans for _collection literals_.
+
+For collection literals that include a _spread operator_, the length of the resulting span is not known at compile time.
+For those cases, we will choose a maximum length for stack allocation and generate code that falls back to heap allocation at runtime if that length is exceeded.
+As per the _collection literals_ working group, a hidden diagnostic will be reported in cases where stack allocation may not be possible at runtime.
+
+For instance, the expression `(ReadOnlySpan<int>)[..e]` could be emitted as:
+```csharp
+var _tmp1 = new FixedSizeArray8<int>();
+int[] _tmp2 = Enumerable.TryGetNonEnumeratedCount(e, out int n) && n <= 8
+    ? _tmp1.Array
+    : new int[n];
+
+int i = 0;
+foreach (var item in e)
+    _tmp2[i++] = item;
+
+var _tmp3 = new ReadOnlySpan<int>(_tmp2, 0, n);
+```
+
+## Example
+Consider the following method with multiple calls to a `params` method:
+```csharp
+// static void WriteLine(string format, params ReadOnlySpan<object> args);
+
+static void WriteDictionary<K, V>(Dictionary<K, V> dictionary)
+{
+    WriteLine("Dictionary");
+
+    foreach (var (k, v) in dictionary)
+        WriteLine("{0}, {1}", k, v);
+
+    WriteLine("Count = {0}", dictionary.Count);
+}
+```
+
+The method could be lowered to:
+```csharp
+static void WriteDictionary<K, V>(Dictionary<K, V> dictionary)
+{
+    FixedSizeArray3 _tmp1 = new FixedSizeArray3();
```

Comment:
```suggestion
    FixedSizeArray2 _tmp1 = new FixedSizeArray2();
```

Only two arguments are used here so seems like this is the more appropriate type